snorkel/map/core.py

```diff
@@ -0,0 +1,115 @@
+import inspect
+from enum import Enum, auto
+from types import SimpleNamespace
+from typing import Any, Callable, Mapping, NamedTuple, Optional, Union
+
+from snorkel.types import DataPoint, FieldMap
+
+
+class MapperMode(Enum):
+    NONE = auto()
+    NAMESPACE = auto()
+    PANDAS = auto()
+    DASK = auto()
+    SPARK = auto()
+
+
+def namespace_to_dict(x: Union[SimpleNamespace, NamedTuple]) -> dict:
+    """Convert a SimpleNamespace or NamedTuple to a dict"""
+    if isinstance(x, SimpleNamespace):
+        return vars(x)
+    return x._asdict()
+
+
+class Mapper:
+    def __init__(
+        self,
+        field_names: Mapping[str, str],
+        mapped_field_names: Optional[Mapping[str, str]] = None,
+    ) -> None:
+        """Map data points to new data points by transforming, adding
+        additional information, or decomposing into primitives. A Mapper
+        maps an data point to a new data point, possibly with a different
+        schema. Subclasses of Mapper need to implement the `run(...)`
+        method, which takes fields of the data point as input
+        and outputs new fields for the mapped data point.
+        For an example of a Mapper, see
+            `snorkel.labeling.preprocess.nlp.SpacyPreprocessor`
+        Args:
+            * field_names: a map from attribute names of the incoming
+                data points to the input argument names of the
+                `run(...)` method
+            * mapped_field_names: a map from output keys of the
+                `run(...)` method to attribute names of the
+                output data points. If None, the original output
+                keys are used.
+        """
+        self.field_names = field_names
+        self.mapped_field_names = mapped_field_names
+        self.mode = MapperMode.NONE
+
+    def set_mode(self, mode: MapperMode) -> None:
+        self.mode = mode
+
+    def run(self, **kwargs: Any) -> FieldMap:
+        raise NotImplementedError
```

Comment:
If I'm using a Mapper/Preprocessor subclass like SpacyPreprocessor, then I wouldn't expect to have to go look at the run function definition to see what the field names are. It's clear in the lambda_preprocessor case since I'm defining it on the spot. IMO this could lead to a lot of unexpected behavior.

Reply:
Actually I can add this.